metrics/ci_integrations/test/source/jenkins/adapter/jenkins_source_client_adapter_test.dart

```diff
@@ -0,0 +1,505 @@
+      }).toList();
+    }
+
+    /// Creates a [BuildData] instance with the given [buildNumber]
+    /// and default build arguments.
+    BuildData createBuildData({
+      @required int buildNumber,
+      BuildStatus buildStatus = BuildStatus.successful,
+    }) {
+      return BuildData(
+        buildNumber: buildNumber,
+        startedAt: defaultDateTime,
+        buildStatus: buildStatus,
+        duration: defaultDuration,
+        workflowName: jobName,
+        url: defaultBuildUrl,
+        coverage: defaultCoverage,
+      );
+    }
+
+    setUp(() {
+      reset(jenkinsClientMock);
+      responses.reset();
+    });
+
+    test(
+      ".processJenkinsBuilds() maps builds which are not building",
+      () {
+        final jenkinsBuilds = [
+          createJenkinsBuild(buildNumber: 1),
+          createJenkinsBuild(buildNumber: 2, building: true),
+        ];
+        final expected = [createBuildData(buildNumber: 1)];
+
+        whenFetchArtifact().thenAnswer(responses.artifact);
+
+        final result = adapter.processJenkinsBuilds(jenkinsBuilds, jobName);
+
+        expect(result, completion(equals(expected)));
+      },
+    );
+
+    test(
+      ".processJenkinsBuilds() maps builds satisfying the given startFromBuildNumber",
+      () {
+        final jenkinsBuilds = createJenkinsBuilds(buildNumbers: [1, 2]);
+        final expected = [createBuildData(buildNumber: 2)];
+
+        whenFetchArtifact().thenAnswer(responses.artifact);
+
+        final result = adapter.processJenkinsBuilds(
+          jenkinsBuilds,
+          jobName,
+          startFromBuildNumber: 1,
+        );
+
+        expect(result, completion(equals(expected)));
+      },
+    );
+
+    test(
+      ".processJenkinsBuilds() throws a StateError if fetching an artifact content fails for any of the given builds",
+      () {
+        const fileName = 'coverage-summary.json';
+        const relativePath = 'test/$fileName';
+        final jenkinsBuild = createJenkinsBuild(buildNumber: 1);
+        final nonExistingCoverageJenkinsBuild = createJenkinsBuild(
+          buildNumber: 2,
+          artifacts: const [
+            JenkinsBuildArtifact(
+              fileName: fileName,
+              relativePath: relativePath,
+            ),
+          ],
+        );
+        final jenkinsBuilds = [jenkinsBuild, nonExistingCoverageJenkinsBuild];
+
+        whenFetchArtifact(
+          buildUrlThat: equals(jenkinsBuild.url),
+          relativePathThat: anything,
+        ).thenAnswer(responses.artifact);
+
+        whenFetchArtifact(
+          buildUrlThat: equals(nonExistingCoverageJenkinsBuild.url),
+          relativePathThat: equals(relativePath),
+        ).thenAnswer((_) => responses.error<Map<String, dynamic>>());
+
+        final result = adapter.processJenkinsBuilds(jenkinsBuilds, jobName);
+
+        expect(result, throwsStateError);
+      },
+    );
+
+    test(
+      ".processJenkinsBuilds() fetches a coverage for each build",
```

Comment:
I wanted to start a broader discussion here about making it visible for testing - while we test this method in isolation and much simpler we can't guarantee that it's used in the fetchBuilds & fetchBuildsAfter - so it feels like that we need to test it through those interface methods even if that means more repetitive tests. What do you think?